### ISSUE TITLE
CMake error and warning in Windows

### ISSUE THREAD

**Issue by stefanpsz** [BioMedIA/MIRTK#52]
I'm trying to compile MIRTK in Windows and I get the following error and warning in CMake:

CMake Error at CMake/Settings.cmake:50 (set_property):
  set_property could not find CACHE variable CMAKE_BUILD_TYPE.  Perhaps it
  has not yet been created.
Call Stack (most recent call first):
  CMake/Basis/ProjectTools.cmake:1895 (include)
  CMake/Basis/ProjectTools.cmake:2303 (basis_initialize_settings)
  CMakeLists.txt:74 (basis_project_begin)

CMake Warning at CMakeLists.txt:84 (message):
  MIRTK wrapper script not implemented for Windows

but the process continues until the end. Subsequently pressing "Generate" just replicates the same behaviour and does **not** produce any MSVS Solution as output.

**Comment by stefanpsz** [comment]
I've been able to generate the MSVS solution file by running cmake on the command prompt:

c:\Workspace\Code\C++\MIRTK\Build>cmake . -DCMAKE_BUILD_TYPE=Release

which IMHO, is a workaround, but not really a solution. Nevertheless, I still get a warning:

CMake Warning (dev) at CMake/Basis/TargetTools.cmake:411 (add_dependencies):
  Policy CMP0046 is not set: Error on non-existent dependency in
  add_dependencies.  Run "cmake --help-policy CMP0046" for policy details.
  Use the cmake_policy command to set the policy and suppress this warning.

  The dependency target "mirtk" of target "help-rst" does not exist.
Call Stack (most recent call first):
  CMakeLists.txt:99 (basis_add_dependencies)
This warning is for project developers.  Use -Wno-dev to suppress it.

**Comment by schuhschuh** [comment]
> I've been able to generate the MSVS solution file by running cmake on the command prompt.

I think you might have just needed to press the "Configure" button multiple times until the "Generate" button is no longer greyed out.

> The dependency target "mirtk" of target "help-rst" does not exist.

That is a mistake of course and related to the fact that the "mirtk" wrapper script is currenly implemented in Bash and thus not available on Windows as the earlier Warning said. It should not affect the build of the C++ libraries and commands.

**Comment by stefanpsz** [comment]
The Generate button is never greyed out (It appears clickable since the very beginning, which might be a cmake-gui problem). I tried clicking configure 6 or 7 times, but I get the exact output with the same error every time

**Comment by stefanpsz** [comment]
When compiling these are some of the errors I get:

24>C:\Workspace\Code\C++\MIRTK\Modules\Numerics\src\mirtkArith.cc(69): error C2065: 'M_PI': undeclared identifier
24>C:\Workspace\Code\C++\MIRTK\Modules\Numerics\src\mirtkArith.cc(80): error C2065: 'M_PI': undeclared identifier
24>C:\Workspace\Code\C++\MIRTK\Modules\Numerics\src\mirtkArith.cc(82): error C2065: 'M_PI': undeclared identifier

24>C:\Workspace\Code\C++\MIRTK\Modules\Common\include\mirtkCifstream.h(27): fatal error C1083: Cannot open include file: 'zlib.h': No such file or directory **Even though I have ZLIB_INCLUDE_DIR correctly set (I got the exact same problem when compiling Nifticlib and solved it in a very ugly way)**

27>C:\Workspace\Code\C++\MIRTK\Modules\ImageIO\src\mirtkNiftiImage.h(27): fatal error C1083: Cannot open include file: 'nifti1_io.h': No such file or directory **Nifticlib_INCLUDE_DIR correctly set as well**

29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2039: 'ptrdiff_t': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2873: 'ptrdiff_t': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2039: 'div_t': is not a member of '`global namespace''
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2039: 'ptrdiff_t': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2873: 'div_t': symbol cannot be used in a using-declaration
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2873: 'ptrdiff_t': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2039: 'ldiv_t': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2873: 'ldiv_t': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2039: 'abort': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2873: 'abort': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2039: 'abs': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2873: 'abs': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2039: 'atof': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2873: 'atof': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2039: 'atoi': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2873: 'atoi': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2039: 'atol': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2873: 'atol': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2039: 'bsearch': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2873: 'bsearch': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2039: 'calloc': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2873: 'calloc': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2039: 'div': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2873: 'div': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(22): error C2039: 'exit': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(22): error C2873: 'exit': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(22): error C2039: 'free': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(22): error C2873: 'free': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(23): error C2039: 'labs': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(23): error C2873: 'labs': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(23): error C2039: 'ldiv': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(23): error C2873: 'ldiv': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(23): error C2039: 'malloc': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(23): error C2873: 'malloc': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(24): error C2039: 'mblen': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(24): error C2873: 'mblen': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(24): error C2039: 'mbstowcs': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(24): error C2873: 'mbstowcs': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(24): error C2039: 'mbtowc': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(24): error C2873: 'mbtowc': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(25): error C2039: 'qsort': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(25): error C2873: 'qsort': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(25): error C2039: 'rand': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(25): error C2873: 'rand': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(25): error C2039: 'realloc': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(25): error C2873: 'realloc': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(26): error C2039: 'srand': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(26): error C2873: 'srand': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(26): error C2039: 'strtod': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(26): error C2873: 'strtod': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(26): error C2039: 'strtol': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(26): error C2873: 'strtol': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(27): error C2039: 'strtoul': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(27): error C2873: 'strtoul': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(28): error C2039: 'wcstombs': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(28): error C2873: 'wcstombs': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(28): error C2039: 'wctomb': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(28): error C2873: 'wctomb': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(30): error C2039: 'lldiv_t': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(30): error C2873: 'lldiv_t': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(33): error C2039: 'getenv': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(33): error C2873: 'getenv': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(34): error C2039: 'system': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(34): error C2873: 'system': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(37): error C2039: 'atoll': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(37): error C2873: 'atoll': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(37): error C2039: 'llabs': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(37): error C2873: 'llabs': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(37): error C2039: 'lldiv': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(37): error C2873: 'lldiv': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(38): error C2039: 'strtof': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(38): error C2873: 'strtof': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(38): error C2039: 'strtold': is not a member of '`global namespace''
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2039: 'div_t': is not a member of '`global namespace''
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2873: 'div_t': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(38): error C2873: 'strtold': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(39): error C2039: 'strtoll': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(39): error C2873: 'strtoll': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(39): error C2039: 'strtoull': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(39): error C2873: 'strtoull': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(41): error C2039: '_Exit': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(41): error C2873: '_Exit': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(41): error C2039: 'at_quick_exit': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(41): error C2873: 'at_quick_exit': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(41): error C2039: 'quick_exit': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(41): error C2873: 'quick_exit': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(32): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(32): error C2146: syntax error: missing ';' before identifier 'FILE'
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(32): warning C4091: '': ignored on left of 'int' when no variable is declared
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(36): error C2039: '_Mbstatet': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(36): error C2873: '_Mbstatet': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(38): error C2039: 'fpos_t': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(38): error C2873: 'fpos_t': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(39): error C2039: 'clearerr': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(39): error C2873: 'clearerr': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(39): error C2039: 'fclose': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(39): error C2873: 'fclose': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(39): error C2039: 'feof': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(39): error C2873: 'feof': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(40): error C2039: 'ferror': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(40): error C2873: 'ferror': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(40): error C2039: 'fflush': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(40): error C2873: 'fflush': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(40): error C2039: 'fgetc': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(40): error C2873: 'fgetc': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(41): error C2039: 'fgetpos': is not a member of '`global namespace''
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(41): error C2873: 'fgetpos': symbol cannot be used in a using-declaration
29>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdio(41): fatal error C1003: error count exceeds 100; stopping compilation
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2039: 'ldiv_t': is not a member of '`global namespace''
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2873: 'ldiv_t': symbol cannot be used in a using-declaration
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2039: 'abort': is not a member of '`global namespace''
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2873: 'abort': symbol cannot be used in a using-declaration
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2039: 'abs': is not a member of '`global namespace''
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2873: 'abs': symbol cannot be used in a using-declaration
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2039: 'atof': is not a member of '`global namespace''
28>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2873: 'atof': symbol cannot be used in a using-declaration

**Comment by schuhschuh** [comment]
> Nifticlib_INCLUDE_DIR correctly set as well

How did you check that it is set correctly? What does it point to ? Are you using the included NiftiCLib or do you set ```WITH_NiftiCLib``` to ON ? Can you try with OFF ?

**Comment by schuhschuh** [comment]
> CMake Warning at CMakeLists.txt:84 (message):
> MIRTK wrapper script not implemented for Windows

This warning is now resolved. The ```mirtk``` wrapper script is now implemented in Python. Thus, MIRTK with applications requires now Python.

**Comment by stefanpsz** [comment]
> How did you check that it is set correctly? What does it point to ? Are you using the included NiftiCLib or do you set WITH_NiftiCLib to ON ? Can you try with OFF ?

I've compiled and run the INSTALL MSVS project for Nifticlib with C:/Workspace/libs/NIFTI as target directory. This means the tree is 

`C:/Workspace/libs/NIFTI`
`C:/Workspace/libs/NIFTI/include`
`C:/Workspace/libs/NIFTI/lib`
`C:/Workspace/libs/NIFTI/bin`

so I've set `Nifticlib_INCLUDE_DIR` as `C:/Workspace/libs/NIFTI/include` and `Nifticlib_LIBRARY` as `C:/Workspace/libs/NIFTI/lib/niftiio.lib`

Anyway if you tell me that by setting  `WITH_NiftiCLib` to OFF I will still be able to work with Nifti files I am more than happy to do so.

> This warning is now resolved. The mirtk wrapper script is now implemented in Python. Thus, MIRTK with applications requires now Python.

Thank you!!

**Comment by schuhschuh** [comment]
> Anyway if you tell me that by setting WITH_NiftiCLib to OFF I will still be able to work with Nifti files I am more than happy to do so.

Both should work in the end. See the installation instructions for what this option does. It is to force the use of a system installation of NiftiCLib. Otherwise the one included in MIRTK is used. That is why it is an advanced option.

Are your header files of NiftiCLib in a subdirectory named "nifti" ?

**Comment by schuhschuh** [comment]
I think I see the problem with an external NiftiCLib. Never tried it myself. In Modules/ImageIO/config/Depends.cmake,  add the following right after the ```basis_find_package (NiftiCLib REQUIRED)```:

```
basis_use_package (NiftiCLib)
```

Can you please try this and see if the cannot find include file errors disappear?

**Comment by stefanpsz** [comment]
> Are your header files of NiftiCLib in a subdirectory named "nifti" ?

Yes

> Can you please try this and see if the cannot find include file errors disappear?

Indeed they do. Although (rather obviously I guess) the following errors still remain:

`Cannot open include file: 'zlib.h': No such file or directory`

**Comment by stefanpsz** [comment]
Another missing error I found:

```
3>    File "C:/Workspace/Code/C++/MIRTK/Modules/Image/src/mirtkForEachVoxelFunction.py", line 47
3>      print "usage: " + sys.argv[0] + ' <arity> <file>'
3>                    ^
3>  SyntaxError: Missing parentheses in call to 'print'
3>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code 1.
```

This one is quite weird b/c I cannot really see any syntax error.

**Comment by stefanpsz** [comment]
> This one is quite weird b/c I cannot really see any syntax error.

I'm afraid is actually quite an annoying problem. Apparently in Windows there is no support for non-parenthesis commands

```
Z:\>python test.py hello
  File "test.py", line 3
    print "usage: " + sys.argv[0] + ' <arity> <file> ' + sys.argv[1]
                  ^
SyntaxError: Missing parentheses in call to 'print'
```

Changing the code to `print ("usage: " + sys.argv[0] + ' <arity> <file> ' + sys.argv[1])` yields:

```
Z:\>python test.py hello
usage: test.py <arity> <file> hello
```

**Comment by schuhschuh** [comment]
> I'm afraid is actually quite an annoying problem.

Well, that's ok. There's not so much Python. And it's not related to Windows. That's because you probably have Python 3.

**Comment by schuhschuh** [comment]
> Cannot open include file: 'zlib.h': No such file or directory

That error is not related to NiftiCLib, hence, the previous change did not attempt to fix it. First, I will changed the default for ```WITH_ZLIB``` on Windows to OFF. I was not considering that it is not available on this platform. Eventually, it is probably best to include the zlib source files in the ThirdParty directory.

I don't have access to Windows now (my VM is running, though, and I installed MSVS 2015 Community Edition yesterday). Can you provide some more details about your zlib installation ? From where did you get it, how did you install it, and what is the file hierarchy of the installation ?

**Comment by stefanpsz** [comment]
> From where did you get it?

[http://sourceforge.net/projects/libpng/files/zlib/1.2.8/](http://sourceforge.net/projects/libpng/files/zlib/1.2.8/)

> how did you install it?

1. Run CMake
2. Build the generated ZLIB MSVS solution
3. Installing into  `C:\Workspace\libs\zlib` by building the INSTALL project of the solution

> what is the file hierarchy of the installation ?

`C:\Workspace\libs\zlib`
`C:\Workspace\libs\zlib\bin`
`C:\Workspace\libs\zlib\include`
`C:\Workspace\libs\zlib\lib`
`C:\Workspace\libs\zlib\share`

The files I use in the MIRTK CMake configuration are:

`C:\Workspace\libs\zlib\include\zconf.h`
`C:\Workspace\libs\zlib\include\zlib.h`
`C:\Workspace\libs\zlib\lib\zlibstaticd.lib`

i.e.

`ZLIB_INCLUDE_DIR=C:/Workspace/libs/zlib/include`
`ZLIB_LIBRARY_DEBUG=C:\Workspace\libs\zlib\lib\zlibstaticd.lib`
`ZLIB_LIBRARY_RELEASE=C:\Workspace\libs\zlib\lib\zlibstaticd.lib`

**Comment by schuhschuh** [comment]
Thanks

**Comment by schuhschuh** [comment]
> 24>C:\Workspace\Code\C++\MIRTK\Modules\Numerics\src\mirtkArith.cc(69): error C2065: 'M_PI': undeclared identifier

This is now fixed by introducing our own constant ```mirtk::pi``` which is set to ```M_PI``` in ```mirtkMath.cc``` where the ```cmath``` header is included after ```#define _USE_MATH_DEFINES``` before any other include.

**Comment by schuhschuh** [comment]
@stefanpsz For the error relating to the STL symbols not being in the global namespace, can you post for one of them a more comprehensive summary of all MSVC error messages? Such as which compilation unit (.cc file) was being built, and the include hierarchy that ends up at the error location?

**Comment by stefanpsz** [comment]
> @stefanpsz For the errors relating to the STL symbols not being in the global namespace, can you post for one of them a more comprehensive summary of all MSVC error messages? Such as which compilation unit (.cc file) was being built, and the include hierarchy that ends up at the error location?

The affected files seem to be all application files such as:

```
calculate.cc
calculate-exponential-map.cc
calculate-logarithmic-map.cc
close-image.cc
concatenate-images.cc
convert-image.cc
cut-brain.cc
```

With a bit of [googling](http://arstechnica.com/civis/viewtopic.php?f=20&t=322185) I found out that the most probable problem is that `using namespace std;` is missing. I remember there was a bit of discussion about whether this line should be avoided and that `std::` should be prefixed everywhere. I do not know what happened with this in the end, but if you think this is not the problem let me know.

**Comment by schuhschuh** [comment]
> using namespace std:: is missing.

Well, while that may fix it, it's not the real solution. I was looking for a complete compiler error backtrace for one of the error messages. Otherwise I cannot trace the origin of the error. A general list of where it appears is not useful. Can you copy and paste the direct compiler output ? Not only the error summary of MSVS ?

**Comment by schuhschuh** [comment]
> ```std::``` should be prefixed everywhere

That's what could be done. However, as I dislike to have it all over the place, I explicitly "import" all needed STL symbols into the ```mirtk``` namespace via ```using``` keyword (not ```using namespace```). Either some of the needed symbols are missing, or something else is going on.

**Comment by stefanpsz** [comment]
> and the include hierarchy that ends up at the error location?

Here it is, let me know if this is what you need

```
15>  Note: including file:  C:\Workspace\Code\C++\MIRTK\Modules\Common\include\mirtkPair.h
15>  Note: including file:   C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\utility
15>  Note: including file:    C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xstddef
15>  Note: including file:     C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef
15>  Note: including file:      C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\yvals.h
15>  Note: including file:       C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xkeycheck.h
15>  Note: including file:       C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\crtdefs.h
15>  Note: including file:        C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\vcruntime.h
15>  Note: including file:       C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\use_ansi.h
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2039: 'ptrdiff_t': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2873: 'ptrdiff_t': symbol cannot be used in a using-declaration
15>  Note: including file:     C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2039: 'div_t': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2873: 'div_t': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2039: 'ldiv_t': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(18): error C2873: 'ldiv_t': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2039: 'abort': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2873: 'abort': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2039: 'abs': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(19): error C2873: 'abs': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2039: 'atof': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2873: 'atof': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2039: 'atoi': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2873: 'atoi': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2039: 'atol': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(20): error C2873: 'atol': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2039: 'bsearch': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2873: 'bsearch': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2039: 'calloc': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2873: 'calloc': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2039: 'div': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(21): error C2873: 'div': symbol cannot be used in a using-declaration
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(22): error C2039: 'exit': is not a member of '`global namespace''
15>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstdlib(22): error C2873: 'exit': symbol cannot be used in a using-declaration
... etc
```

**Comment by schuhschuh** [comment]
Note: Need to document that at least Visual Studio 2013 is required to build the MIRTK on Windows because older versions do not support the C++ type aliases yet.

**Comment by schuhschuh** [comment]
The current code does not use any ```__declspec``` in the declarations of its API classes and functions. The shared library build fails therefore because no ```.lib``` files are created.

- [ ] Export shared library symbols from DLL.

**Comment by schuhschuh** [comment]
There is some CMake functionality that can help with the export:
- Instead of writing the export/import macros for each MIRTK module ourselves, we can use the [GenerateExportHeader](https://cmake.org/cmake/help/v2.8.12/cmake.html#module:GenerateExportHeader) module function.
- CMake 3.4 introduced the target property [WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/v3.4/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) which will export all symbols except static data members by auto-generating a ```.def``` file with the list of all symbols extracted from the ```.obj``` files which is passed on to the linker. This does not require adding a ```<Lib>_EXPORT``` macro to all class and function declarations. As Windows users can easily install the latest CMake version by downloading the installer from https://cmake.org/download/, I am fine with requiring at least this version. For Ubuntu 14 users, we still should only require version 2.8.12. See also [this Kitware blog post](https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/) on more information on the new ```WINDOWS_EXPORT_ALL_SYMBOLS``` feature.

**Comment by schuhschuh** [comment]
The error messages
```
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2039: 'ptrdiff_t': is not a member of '\`global namespace''
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\cstddef(18): error C2873: 'ptrdiff_t': symbol cannot be used in a using-declaration
```
are caused by the include of ```windows.h```. As we don't use any Windows specific types and functions but only the STL, we don't actually need to include that header file. The ```mirtkWindows.h``` file in the Common module is then obsolete (also considering that a recent MSVC compiler with snprintf is required).

**Comment by ghisvail** [comment]
Plus `windows.h` is notorious for pulling a lot of unwanted stuff. We would be better off without it or only *selectively* included in implementation files that require it.
